packages/server/src/runtime/builtin/extract.rs

```diff
@@ -122,51 +125,72 @@ where
-				tg::error!(!source, "failed to get the entry from the archive")
-			})?;
-			let header = entry.header();
-			let path = PathBuf::from(
-				entry
-					.path()
-					.map_err(|source| tg::error!(!source, "failed to get the entry path"))?
-					.as_ref(),
-			);
-			match header.entry_type() {
-				async_tar::EntryType::Regular => {
-					let blob = tg::Blob::with_reader(server, entry.compat()).await?;
-					let file = tg::File::with_contents(blob);
-					let artifact = tg::Artifact::File(file);
-					Ok((path, artifact))
-				},
-				async_tar::EntryType::Symlink => {
-					let target = header
-						.link_name()
-						.map_err(|source| tg::error!(!source, "failed to read the symlink target"))?
-						.ok_or_else(|| tg::error!("no symlink target stored in the archive"))?;
-					let symlink = tg::Symlink::with_target(target.as_ref().into());
-					let artifact = tg::Artifact::Symlink(symlink);
-					Ok((path, artifact))
-				},
-				async_tar::EntryType::Directory => {
-					let directory = tg::Directory::with_entries([].into());
-					let artifact = tg::Artifact::Directory(directory);
-					Ok((path, artifact))
-				},
-				_ => Err(tg::error!("unsupported entry type")),
-			}
-		})
-		.try_collect()
-		.await?;
+		.map_err(|source| tg::error!(!source, "failed to get the entries from the archive"))?;
+
+	// Process the entries.
+	while let Some(entry) = entries.next().await {
+		let entry = entry
+			.map_err(|source| tg::error!(!source, "failed to get the entry from the archive"))?;
+		let header = entry.header();
+		let path = PathBuf::from(
+			entry
+				.path()
+				.map_err(|source| tg::error!(!source, "failed to get the entry path"))?
+				.as_ref(),
+		);
+		match header.entry_type() {
+			async_tar::EntryType::Regular | async_tar::EntryType::Other(_) => {
+				let blob = tg::Blob::with_reader(server, entry.compat()).await?;
+				let file = tg::File::with_contents(blob);
+				let artifact = tg::Artifact::File(file);
+				processed_entries.push((path, artifact));
+			},
+			async_tar::EntryType::Symlink => {
+				let target = header
+					.link_name()
+					.map_err(|source| tg::error!(!source, "failed to read the symlink target"))?
+					.ok_or_else(|| tg::error!("no symlink target stored in the archive"))?;
+				let symlink = tg::Symlink::with_target(target.as_ref().into());
+				let artifact = tg::Artifact::Symlink(symlink);
+				processed_entries.push((path, artifact));
+			},
+			async_tar::EntryType::Directory => {
+				let directory = tg::Directory::with_entries([].into());
+				let artifact = tg::Artifact::Directory(directory);
+				processed_entries.push((path, artifact));
+			},
+			async_tar::EntryType::Link => {
+				let target = header
+					.link_name()
+					.map_err(|source| {
+						tg::error!(!source, "failed to read the hard link target path")
+					})?
+					.ok_or_else(|| tg::error!("no hard link target path stored in the archive"))?;
+				let target: &Path = target.as_ref().into();
```

Comment:
Isn't the target relative to the entry path's parent, not the root of the archive?

Reply:
What I see are paths relative to the archive root (probably just my `$pwd`). When I run with `tar -cf test.tar -C /absolute/path .`, both hardlink and entry paths are stored relatively:
```
hardlink target: "./test.ts"
```

```
./
./test.ts
./link
```

This breaks our extraction because `tg::Directory::builder` expects paths to have at least one component.

Reply:
Just pushed another commit that strips `./` from paths in `tar` extraction, and skips the entry if the resulting path is empty. In my testing, this handles `tar` files archived using `-C` properly. The entry paths and the hardlink targets are stored in the same manner, so we can find the hardlink target.

The `zip` path does not have an equivalent (the closest thing is `--junk-paths`, which simply uses filenames).